### PR TITLE
Add language server for Bebop

### PR DESCRIPTION
**Pull request by patriksvensson**
This PR adds VSCode language server support for Bebop.

## Todo

There is a couple of things that need to be done before it can be used, mainly:

- [x] Need to package the extension together with the LSP server.
- [x] Need to find LSP server without hard-coded paths.
- [x] Need to add mechanism in `SchemaParser` for resolving files referenced by `import` statement.
- [x] Removing documents from the buffer when they are closed.
- [x] Add startup progress for language server.
- [x] Cleaning up in general.

A lot of things would be nicer and cleaner if the Bebop parser exposed an AST, 
(and open up for a lot of cool features) but I've done my best abusing the `Tokenizer`, and I think it's OK.

## To try this out
Right now, the path to the language server is hardcoded in `extension.ts` (I will, of course, fix this).

To try the language server, open the `vscode-bebop` folder in VSCode.

1. Run `npm installed` followed by `npm run esbuild` in the `vscode-bebop` folder.
2. Ensure that the solution has been built in debug mode. 
3. Press F5 to open a new VSCode instance with the Language Server loaded.

## Screenshots
Bebop

Bebop1

### CODE REVIEW

**Comment by lynn** [comment]
This looks excellent so far, thank you for working on this!

> A lot of things would be nicer and cleaner if the Bebop parser exposed an AST,

This is something I'm interested in doing right. Currently `BebopSchema` is kind of like a "flattened" AST (to answer your comment, its `Definitions` also contains nested definitions; the `Parent` method on `Definition` can be used to recover the structure to some extent). Which parts of your code would be simplified away if we exposed a real AST?

**Comment by AndrewMD5** [comment]
I haven't had a chance to deeply review this yet but my initial thought is that the language server can be inside bebopc itself and started when a commandline flag is present. this avoids the need to ship multiple binaries and we can leverage current tools for installing the compiler

**Comment by patriksvensson** [comment]
@lynn

> This is something I'm interested in doing right. Currently, BebopSchema is kind of like a "flattened" AST (to answer your comment, its Definitions also contain nested definitions; the Parent method on Definition can be used to recover the structure to some extent). Which parts of your code would be simplified away if we exposed a real AST?

There are a couple of reasons. I'm using semantic highlighting in LSP for syntax highlighting, and I think it would help find ambiguities when deciding what kind of symbol something is. All primitive types are hardcoded in the language server, for example.

It would also open up for things like information when hovering over a token, "Go to definition", "Go to reference", "Rename", and other refactorings.

I also noticed that tokens don't distinguish between a lexeme and the corresponding token value,
which made it difficult to highlight comments and strings.

i.e. `/* Hello World */` token would have a span of `4-14` with a lexeme of `Hello World`, instead of a
token with a span of `1-17` with the lexeme `/* Hello World */` and a value `Hello World`. I added a TextMate grammar
to take care of this though, so not a big problem.

@AndrewMD5

Excellent idea, which will make distribution easier. I'll add it to the compiler, and the extension can try to launch
it as a global tool? If the extension can't find it, we can just tell the user to install the compiler. Did I understand that correctly?

**Comment by AndrewMD5** [comment]
> Excellent idea, which will make distribution easier. I'll add it to the compiler, and the extension can try to launch it as a global tool? If the extension can't find it, we can just tell the user to install the compiler. Did I understand that correctly?

Each release of Bebop ships with two packages; the “bebop” package which contains the runtime for a specific language, and “bebop-tools” which contains the compiler.

We use synced global versioning for each component— version 2.3.0 of the runtime package is guaranteed to be compatible with version 2.3.0 of the compiler which ships inside the corresponding “bebop-tools” package of the same version.

We can setup the VS Code extension so that when it’s built for release we bundle the relevant compiler with it so it’s always present

**Comment by patriksvensson** [comment]
@AndrewMD5 I see. Thanks for the explanation.

Let me take a look at the outstanding issues that aren't related to packaging and perhaps we can revisit this once things are in a shippable state?

**Comment by patriksvensson** [comment]
@AndrewMD5 @lynn Forgot to write in my last message; could you approve the workflows for this PR to better understand the state things are in? Thanks!

**Comment by AndrewMD5** [comment]
> @AndrewMD5 @lynn Forgot to write in my last message; could you approve the workflows for this PR to better understand the state things are in? Thanks!

Done

**Comment by patriksvensson** [comment]
@lynn @AndrewMD5 I've added a mechanism to resolve imported files in `SchemaReader` which I would be interested in getting your feedback on since it involves a change to the parser. The commit is https://github.com/RainwayApp/bebop/pull/197/commits/1876a7106e8ecdd80375d4f6197da07eae7f1329 and the change is at https://github.com/RainwayApp/bebop/pull/197/commits/1876a7106e8ecdd80375d4f6197da07eae7f1329#diff-82f7aa3fdad0b022c52dc4528683526f97e98cd0b5fea2cc3e6db4489941e60a.

Other than that it seems to work fine.

imports

**Comment by patriksvensson** [comment]
@lynn @AndrewMD5 

I will clean up the code tomorrow and start with the packaging.

I noticed that you do most build automation stuff directly in GitHub Actions YAML, but would you be open to a local build script to do the packaging (which would be called from the GH Action)? It would simplify development a lot.

**Comment by AndrewMD5** [comment]
> @lynn @AndrewMD5
> 
> I will clean up the code tomorrow and start with the packaging.
> 
> I noticed that you do most build automation stuff directly in GitHub Actions YAML, but would you be open to a local build script to do the packaging (which would be called from the GH Action)? It would simplify development a lot.

We can execute a script for now and modify it as needed

**Comment by patriksvensson** [comment]
The extension should now work without the hard-coded path, which is one more thing that had to be done before packaging the extension.

**Comment by AndrewMD5** [comment]
> The extension should now work without the hard-coded path, which is one more thing that had to be done before packaging the extension.

FYI
> Ensure .NET Runtime is installed before launching extension.

No need for this. We AOT compile the binary.

File: vscode-bebop/src/extension.ts [hunk 1]
```diff
@@ -46,12 +47,41 @@ export function activate(context: vscode.ExtensionContext) {
...
+
+        let os = platform();
+        if (os === "win32") {
+            return path.resolve(path.join(context.extensionPath, '../bin/compiler/Windows-Debug/bebopc.exe'));
+        } else {
+            return path.resolve(path.join(context.extensionPath, '../bin/compiler/Linux-Debug/bebopc.exe'));
+        }
+    }
+
+    // Use the packaged compiler
```
Comment (AndrewMD5) [inline vscode-bebop/src/extension.ts:65]: bebopc ships for macOS, Linux, and Windows so we’ll need to update the appropriate pathing.

**Comment by patriksvensson** [comment]
@AndrewMD5 @lynn Could someone approve the CI flow for this PR? I will need to run this several times to get the packaging right, so if you could grant me access without you having to approve each run, it would be great!

**Comment by patriksvensson** [comment]
@AndrewMD5 @lynn Nevermind that. I can run them in my fork when I think about it.

**Comment by patriksvensson** [comment]
@AndrewMD5 Good news and bad news. 

The good news is that we now package the VSCode extension.

The bad news is that the LSP server is completely incompatible with assembly trimming. It relies a lot on reflection to hook things up behind the scenes, which results in us getting thousands of trimming warnings and type initializer exceptions at runtime.

I think we will need to reconsider how to distribute the language server. We could install the .NET runtime as part of the extension at the first run and include the compiler as part of the language server. This is the approach that other compiler extensions such as [Bicep](https://github.com/Azure/bicep/blob/main/src/vscode-bicep/src/language/client.ts#L164) take.

**Comment by AndrewMD5** [comment]
> @AndrewMD5 Good news and bad news.
> 
> The good news is that we now package the VSCode extension.
> 
> The bad news is that the LSP server is completely incompatible with assembly trimming. It relies a lot on reflection to hook things up behind the scenes, which results in us getting thousands of trimming warnings and type initializer exceptions at runtime. I should have tested this properly, but never got around to it until it was time for packaging of the extension.
> 
> I think we will need to reconsider how to distribute the language server. We could install the .NET runtime as part of the extension at the first run and include the compiler as part of the language server. This is the approach that other compiler extensions such as [Bicep](https://github.com/Azure/bicep/blob/main/src/vscode-bicep/src/language/client.ts#L164) take.

Let me play around with the AOT compilation. It is using a very old target and its becoming stable in .NET 7, so I'll retarget this branch to see if I can get it working.

**Comment by patriksvensson** [comment]
@AndrewMD5 Any luck with this?

**Comment by AndrewMD5** [comment]
> @AndrewMD5 Any luck with this?

The latest master branch should produce builds that work now

**Comment by patriksvensson** [comment]
Awesome! I'll take a look at the current state of it all tonight when I'm at a computer. 👍

**Comment by AndrewMD5** [comment]
> Awesome! I'll take a look at the current state of it all tonight when I'm at a computer. 👍

Let me know if you have any issues! Master now produces builds for x64 and ARM64 and I’ve updated the resolving logic for the binaries accordingly.

**Comment by patriksvensson** [comment]
@AndrewMD5 Taking a look at it now. I will report back after I've rebased everything against the master branch.

**Comment by patriksvensson** [comment]
@AndrewMD5 It's working! 

1. Download the `vscode` artifact from https://github.com/patriksvensson-forks/bebop/actions/runs/2693110403#artifact and unzip it
2. Open VSCode, press `CTRL+SHIFT+P`, and bring up the `Extensions: Install from VSIX...` palette item
3. Select the VSIX file that you unzipped
4. Open any file with a `.bop` extension
5. Profit

I still got some stuff I need to do, like cleaning up the code base a bit, fixing the VS extension, and some other stuff, but finally getting somewhere.

**Comment by patriksvensson** [comment]
@AndrewMD5 I think I can bikeshed things in the extension for an eternity, so let's consider this PR done for now and ready for review. Happy to get back and make the extension even better once we have something in place.

File: Compiler/LangServer/BebopDiagnosticPublisher.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            });
+        }
+
+        public void PublishDiagnostics(DocumentUri uri, int? version, IEnumerable<BebopDiagnostic> diagnostics)
+        {
+            // Group diagnostics by their filename
+            foreach (var group in diagnostics.GroupBy(x => x.Error.Span.FileName))
+            {
+                // If the schema is `unknown`, assume the diagnostics come from the currently opened file
+                var noSchema = group.Key == "(unknown)";
```
Comment (lynn) [inline Compiler/LangServer/BebopDiagnosticPublisher.cs:41]: Out of curiosity where does this string `"(unknown)"` come from?

File: Compiler/LangServer/Text/Buffer.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        }
+
+        public Range GetRange(Span span)
+        {
+            return new Range
+            {
+                Start = new Position(span.StartLine, span.StartColumn),
+                End = new Position(span.EndLine, span.EndColumn),
+            };
+        }
```
Comment (lynn) [inline Compiler/LangServer/Text/Buffer.cs:33]: Is this redundant to the definition in [Compiler/LangServer/Extensions/SpanExtensions.cs](https://github.com/RainwayApp/bebop/pull/197/files#diff-f0e26fca888de9503511986f9e50d8a6268a475b928c95630b1cbb203141c27c)?

**Review by lynn — APPROVED** [review]
Looks great to me! Thank you for working on this. I tested it locally and it works like a charm.

File: Compiler/LangServer/BebopDiagnosticPublisher.cs (hunk 2 above)
Comment (patriksvensson) [reply Compiler/LangServer/BebopDiagnosticPublisher.cs:41]: It seems to come from the schema reader: https://github.com/RainwayApp/bebop/blob/master/Core/IO/SchemaReader.cs#L34

File: Compiler/LangServer/Text/Buffer.cs (hunk 3 above)
Comment (patriksvensson) [reply Compiler/LangServer/Text/Buffer.cs:33]: Yes, will fix this 👍

**Comment by patriksvensson** [comment]
@lynn @AndrewMD5 I've cleaned up the PR (removed unused methods and similar), responded to code review and squashed the commits.

**Comment by AndrewMD5** [comment]
@patriksvensson thank you for the contribution! sorry for the delays on our end. I'll merge this in and do the work necessary to get it fully released with v3.0.0

**Pull request merged**